doc/man/man8/strvf.8.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man8/strvf.8.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH STRVF 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.\"
.\"
.SH NAME
.B strvf
\- \fISTREAMS\fR verification tool
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBstrvf\fR [\fIoptions\fR]
.PD 0
.HP 12
\fBstrvf\fR {\fB-h\fR|\fB--help\fR}
.HP 12
\fBstrvf\fR {\fB-V\fR|\fB--version\fR}
.HP 12
\fBstrvf\fR {\fB-C\fR|\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
The
.B strvf
command executes a series of subcommands that verify whether or not
.I STREAMS
is currently installed and configured on your system.  All output is sent to
.IR stdout .
Verbose output is always sent to the log file
.BR /var/log/streams/strvf.log .
.\".BR /var/adm/streams/strvf.log .
.PP
These subcommands make sure that the
.I STREAMS
kernel daemons are running and that
.BR open (2s),
.BR putmsg (2s),
.BR getmsg (2s),
.BR ioctl (2s)
and
.BR close (2s)
can be performed on
.BR /dev/echo .
.PP
.B strvf
checks the following items:
.IP \(bu \w'\(bu\(em'u
STREAMS kernel daemons are running.
.PD 0
.IP \(bu
The echo driver (a core STREAMS driver) can be opened.
.IP \(bu
A
.BR putmsg (2s)
can be performed on the echo driver.
.IP \(bu
A
.BR getmsg (2s)
receives the same message sent by
.BR putmsg (2s).
.IP \(bu
A STREAMS ioctl can be passed to the echo driver and acknowledged.
.IP \(bu
The echo driver can be closed.
.PD
.\"
.\"
.SH OPTIONS
.PP
.TP
\fB-n\fR, \fB--nologging\fR
Suppress logging, send output to
.B stdout
and errors to
.BR stderr .
Default: off.
.TP
\fB-b\fR, \fB--basename\fR \fIBASENAME\fR
Only significant when logging.
Set the base name of logging and error files.  Default:
.RB \(lq strvf \(rq.
.TP
\fB-O\fR, \fB--directory\fR \fIDIRECTORY\fR
Only significant when logging.
Set the directory to which log and error files are written.  Default:
.RB \(lq /var/log/streams \(rq.
.TP
\fB-o\fR, \fB--outfile\fR \fIOUTFILE\fR
Only significant when logging.
Set the output file name for logging.  Default:
\(lq\fIBASENAME\fB.log\fR\(rq.
.TP
\fB-e\fR, \fB--errfile\fR \fIERRFILE\fR
Only significant when logging.
Set the output file name for errors.  Default:
\(lq\fIBASENAME\fB.err\fR\(rq.
.TP
\fB-D\fR, \fB--devname\fR \fIDEVNAME\fR
Only significant when logging.
Set the device name to perform test upon.  Default:
.RB \(lq /dev/echo \(rq.
.TP
\fB-q\fR, \fB--quiet\fR
Suppress normal output.
.TP
\fB-d\fR, \fB--debug\fR [\fIlevel\fR]
Increase, or specify the debugging of the output.
.TP
\fB-v\fR, \fB--verbose\fR [\fIlevel\fR]
Increase, or specify the verbosity of the output.
.TP
\fB-h\fR, \fB--help\fR
Print usage information and exit.
.TP
\fB-V\fR, \fB--version\fR
Print version information and exit.
.TP
\fB-C\fR, \fB--copying\fR
Print copying information and exit.
.\"
.\"
.SH DIAGNOSTICS
.PP
.B strvf
returns a zero (0) exit status on success.  If
.B strvf
fails, it returns an exit status of one (1) and prints a diagnostic error
message to
.BR stderr .
.PP
An exit status of zero (0) indicates that
.B strvf
was successful; one (1) indicates that an error occurred and diagnostic message
is printed on
.BR stderr ;
two (2) indicates that the option syntax was in error and a diagnostic message
is printed on
.BR stderr .
.PP
Errors may result from the failure of any of the test in the verification
procedure, failure to open the
.B /dev/echo
device, or inadequate permissions.
.\"
.\"
.SH NOTICES
.PP
The
.B strvf
command does not really accomplish much.  The Linux Fast-STREAMS conformance
verification test suite,
.BR test-streams (8)
is far more exhaustive and yields usable results.  The
.B strvf
command is provided solely for compatibility with
.IR HP-UX \(rg.
.\"
.\"
.SH EXAMPLES
.PP
Sample output:
.RS
.nf
\fC\s-1
.\"-> Logging results to /var/adm/streams/strvf.log
-> Logging results to /var/log/streams/strvf.log
--------------------------------------------------
-> Verify Linux Fast-STREAMS installation.
-> Verify open, putmsg, getmsg, ioctl, and close
-> can be performed on a stream.
--------------------------------------------------
--------------------------------------------------
-> Linux Fast-STREAMS is installed and operational
--------------------------------------------------
\s+1\fP
.fi
.RE
.\"
.\"
.SH DEVICES
.PP
.B /dev/echo
Loopback
.I STREAMS
driver used by
.BR strvf .
.\"
.\"
.SH FILES
.PP
.TP
.B /usr/sbin/strvf
contains the
.B strvf
command.
.TP
.B /var/log/streams/strvf.log
.\".B /var/adm/streams/strvf.log
contains a verbose description and status of all sub-commands.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR echo (4),
.BR test-streams (8).
.\"
.\"
.SH BUGS
.PP
.B strvf
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.IR HP-UX \(rg.
This is an
.IR HP-UX \(rg-specific
command.
.[
hpuxspg
.]
.B strvf
is provided in
.I @PACKAGE_TITLE@
for compatibility with
.IR HP-UX \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IR HP-UX \(rg
does not document any long options.
.IP \(em
.IR HP-UX \(rg
does not document the
.BR -d ", " -q ", " -v ", " -h ", " -V " or " -C options.
.IP \(em
Options
.BR --help " and " --version
are provided for compatibility with GNU coding standards;
.BR --copying ,
OpenSS7 coding standards.
.IP \(em
The format of the output from the
.B strvf
command may be different from that of
.IR HP-UX \(rg.
.\"
.\"
.SH CONFORMANCE
.PP
.IR HP-UX \(rg
documentation.
.[
hpuxdoc
.]
.\"
.\"
.SH HISTORY
.PP
.B strvf
first appeared in
.IR HP-UX \(rg.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/util/strvf.c
/*****************************************************************************

 @(#) File: src/util/strvf.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/util/strvf.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE 600
#endif

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <memory.h>
#include <fcntl.h>
#include <errno.h>
#include <limits.h>

#ifdef _GNU_SOURCE
#include <getopt.h>
#endif

#include <stropts.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/sysmacros.h>
#include <sys/poll.h>
#include <sys/time.h>
#include <sys/ioctl.h>
#include <time.h>
#include <signal.h>
#include <syslog.h>
#include <sys/utsname.h>
#include <sys/strlog.h>

static int logging = 1;
static int debug = 0;			/* default no debug */
static int output = 1;			/* default normal output */

char outfile[256] = "";
char errfile[256] = "";
char outpath[256] = "";
char errpath[256] = "";
char basname[256] = "";
char outpdir[256] = "/var/log/streams";
char devname[256] = "";

static void
copying(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
--------------------------------------------------------------------------------\n\
%1$s\n\
--------------------------------------------------------------------------------\n\
Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>\n\
Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>\n\
Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>\n\
\n\
All Rights Reserved.\n\
--------------------------------------------------------------------------------\n\
This program is free software: you can  redistribute it  and/or modify  it under\n\
the terms of the  GNU Affero  General  Public  License  as published by the Free\n\
Software Foundation, version 3 of the license.\n\
\n\
This program is distributed in the hope that it will  be useful, but WITHOUT ANY\n\
WARRANTY; without even  the implied warranty of MERCHANTABILITY or FITNESS FOR A\n\
PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.\n\
\n\
You should have received a copy of the  GNU Affero General Public License  along\n\
with this program.   If not, see <http://www.gnu.org/licenses/>, or write to the\n\
Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\
--------------------------------------------------------------------------------\n\
U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on behalf\n\
of the U.S. Government (\"Government\"), the following provisions apply to you. If\n\
the Software is supplied by the Department of Defense (\"DoD\"), it is classified\n\
as \"Commercial  Computer  Software\"  under  paragraph  252.227-7014  of the  DoD\n\
Supplement  to the  Federal Acquisition Regulations  (\"DFARS\") (or any successor\n\
regulations) and the  Government  is acquiring  only the  license rights granted\n\
herein (the license rights customarily provided to non-Government users). If the\n\
Software is supplied to any unit or agency of the Government  other than DoD, it\n\
is  classified as  \"Restricted Computer Software\" and the Government's rights in\n\
the Software  are defined  in  paragraph 52.227-19  of the  Federal  Acquisition\n\
Regulations (\"FAR\")  (or any successor regulations) or, in the cases of NASA, in\n\
paragraph  18.52.227-86 of  the  NASA  Supplement  to the FAR (or any  successor\n\
regulations).\n\
--------------------------------------------------------------------------------\n\
Commercial  licensing  and  support of this  software is  available from OpenSS7\n\
Corporation at a fee.  See http://www.openss7.com/\n\
--------------------------------------------------------------------------------\n\
", ident);
}

static void
version(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
%1$s (OpenSS7 %2$s) %3$s (%4$s)\n\
Written by Brian Bidulock.\n\
\n\
Copyright (c) 2008, 2009, 2010, 2015  Monavacon Limited.\n\
Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008  OpenSS7 Corporation.\n\
Copyright (c) 1997, 1998, 1999, 2000, 2001  Brian F. G. Bidulock.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n\
\n\
Distributed by OpenSS7 under GNU Affero General Public License Version 3,\n\
with conditions, incorporated herein by reference.\n\
\n\
See `%1$s --copying' for copying permissions.\n\
", NAME, PACKAGE, VERSION, PACKAGE_ENVR " " PACKAGE_DATE);
}

static void
usage(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stderr, "\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
", argv[0]);
}

static void
help(int argc, char *argv[])
{
	if (!output && !debug)
		return;
	(void) fprintf(stdout, "\
Usage:\n\
    %1$s [options]\n\
    %1$s {-h|--help}\n\
    %1$s {-V|--version}\n\
    %1$s {-C|--copying}\n\
Arguments:\n\
    None.\n\
Options:\n\
    -n, --nologging\n\
        do not log, send output to stdout and stderr\n\
    -b, --basename BASENAME\n\
        file basename, default: 'strvf'\n\
    -O, --directory DIRECTORY\n\
        redirect output to files in DIRECTORY, default: /var/log/streams\n\
    -o, --outfile OUTFILE\n\
        redirect output to OUTFILE, default: ${BASENAME}.log\n\
    -e, --errfile ERRFILE\n\
        redirect errors to ERRFILE, default: ${BASENAME}.err\n\
    -D, --devname DEVNAME\n\
        test device to open, default: '/dev/streams/clone/echo'\n\
    -q, --quiet\n\
        suppress normal output (equivalent to --verbose=0)\n\
    -d, --debug [LEVEL]\n\
        increment or set debug LEVEL [default: 0]\n\
    -v, --verbose [LEVEL]\n\
        increment or set output verbosity LEVEL [default: 1]\n\
        this option may be repeated.\n\
    -h, --help, -?, --?\n\
        print this usage information and exit\n\
    -V, --version\n\
        print version and exit\n\
    -C, --copying\n\
        print copying permission and exit\n\
", argv[0]);
}

static void
strvf_exit(int retval)
{
	if (retval) {
		fprintf(stderr, "--------------------------------------------------\n");
		fprintf(stderr, "-> Linux Fast-STREAMS is NOT operational.\n");
		fprintf(stderr, "--------------------------------------------------\n");
	} else {
		fprintf(stderr, "--------------------------------------------------\n");
		fprintf(stderr, "-> Linux Fast-STREAMS is installed and operational\n");
		fprintf(stderr, "--------------------------------------------------\n");
	}
	fflush(stderr);
	exit(retval);
}

static void
strvf(int argc, char *argv[])
{
	int fd;

	if (basname[0] == '\0')
		snprintf(basname, sizeof(basname), "strvf");
	if (logging || outfile[0] != '\0') {
		if (outfile[0] == '\0')
			snprintf(outfile, sizeof(outfile), "%s.log", basname);
		snprintf(outpath, sizeof(outpath), "%s/%s", outpdir, outfile);
		if (output > 1)
			syslog(LOG_NOTICE, "Redirecting stdout to file %s", outpath);
		fflush(stdout);
		if (freopen(outpath, "a", stderr) == NULL) {
			syslog(LOG_ERR, "%m");
			syslog(LOG_ERR, "Could not redirect stdout to %s", outpath);
			exit(1);
		}
	}
	if (logging || errfile[0] != '\0') {
		if (errfile[0] == '\0')
			snprintf(errfile, sizeof(errfile), "%s.err", basname);
		snprintf(errpath, sizeof(errpath), "%s/%s", outpdir, errfile);
		if (output > 1)
			syslog(LOG_NOTICE, "Redirecting stderr to file %s", errpath);
		fflush(stderr);
		if (freopen(errpath, "a", stderr) == NULL) {
			syslog(LOG_ERR, "%m");
			syslog(LOG_ERR, "Could not redirect stderr to %s", errpath);
			exit(1);
		}
	}
	syslog(LOG_NOTICE, "Startup complete.");
	fprintf(stderr, "-> Logging results to %s\n", outpath);
	fprintf(stderr, "--------------------------------------------------\n");
	fprintf(stderr, "-> Verify Linux Fast-STREAMS installation.\n");
	fprintf(stderr, "-> Verify open, putmsg, getmsg, ioctl, and close\n");
	fprintf(stderr, "-> can be performed on a stream.\n");
	fprintf(stderr, "--------------------------------------------------\n");
	fflush(stderr);
	{
		fprintf(stdout, "Open device %s\n", devname);
		fflush(stdout);
		if ((fd = open(devname, O_RDWR)) < 0) {
			fprintf(stdout, "Open failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Open successful.\n");
		fflush(stdout);
	}
	{
		int ret, flags = 0;
		char ctlbuf[] = "\0\1\2\3\4\5\6\7";
		char datbuf[] = "\0\1\2\3\4\5\6\7";
		struct strbuf ctl = { sizeof(ctlbuf), sizeof(ctlbuf), ctlbuf };
		struct strbuf dat = { sizeof(datbuf), sizeof(datbuf), datbuf };

		fprintf(stdout, "Putmsg to device %s\n", devname);
		fflush(stdout);
		if ((ret = putmsg(fd, &ctl, &dat, flags)) < 0) {
			fprintf(stdout, "Putmsg failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Putmsg successful.\n");
		fflush(stdout);
	}
	{
		int ret, flags = 0;
		char ctlbuf[8];
		char datbuf[8];
		struct strbuf ctl = { sizeof(ctlbuf), sizeof(ctlbuf), ctlbuf };
		struct strbuf dat = { sizeof(datbuf), sizeof(datbuf), datbuf };

		fprintf(stdout, "Getmsg from device %s\n", devname);
		fflush(stdout);
		if ((ret = getmsg(fd, &ctl, &dat, &flags)) < 0) {
			fprintf(stdout, "Getmsg failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Getmsg successful.\n");
		fflush(stdout);
	}
	{
		int ret;
		char iocbuf[] = "\0\1\2\3\4\5\6\7";
		struct strioctl ioc = { I_STR, 0, sizeof(iocbuf), iocbuf };

		fprintf(stdout, "Ioctl of device %s\n", devname);
		fflush(stdout);
		if ((ret = ioctl(fd, I_STR, &ioc)) < 0) {
			fprintf(stdout, "Ioctl failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Ioctl successful.\n");
		fflush(stdout);
	}
	{
		fprintf(stdout, "Close device %s\n", devname);
		fflush(stdout);
		if (close(fd) < 0) {
			fprintf(stdout, "Close failed.\n");
			fflush(stdout);
			perror(__FUNCTION__);
			strvf_exit(1);
		}
		fprintf(stdout, "Close successful.\n");
		fflush(stdout);
	}
	strvf_exit(0);
}

int
main(int argc, char *argv[])
{
	for (;;) {
		int c, val;

#if defined _GNU_SOURCE
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"nologging",	no_argument,		NULL, 'n'},
			{"basename",	required_argument,	NULL, 'b'},
			{"directory",	required_argument,	NULL, 'O'},
			{"outfile",	required_argument,	NULL, 'o'},
			{"errfile",	required_argument,	NULL, 'e'},
			{"devname",	required_argument,	NULL, 'D'},
			{"quiet",	no_argument,		NULL, 'q'},
			{"debug",	optional_argument,	NULL, 'd'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{"?",		no_argument,		NULL, 'H'},
			{ 0, }
		};
		/* *INDENT-ON* */

		c = getopt_long_only(argc, argv, "n:b:O:o:e:D:qd::v::hVC?W:", long_options,
				     &option_index);
#else				/* defined _GNU_SOURCE */
		c = getopt(argc, argv, "n:b:O:o:e:D:qdvhVC?");
#endif				/* defined _GNU_SOURCE */
		if (c == -1) {
			if (debug)
				fprintf(stderr, "%s: done options processing\n", argv[0]);
			break;
		}
		switch (c) {
		case 0:
			goto bad_usage;
		case 'n':	/* -n, --nologging */
			logging = 0;
			break;
		case 'b':
			strncpy(basname, optarg, sizeof(basname));
			break;
		case 'O':
			strncpy(outpdir, optarg, sizeof(outpdir));
			break;
		case 'o':
			strncpy(outfile, optarg, sizeof(outfile));
			break;
		case 'e':
			strncpy(errfile, optarg, sizeof(errfile));
			break;
		case 'D':
			strncpy(devname, optarg, sizeof(devname));
			break;
		case 'd':	/* -d, --debug [LEVEL] */
			if (debug)
				fprintf(stderr, "%s: increasing debug verbosity\n", argv[0]);
			if (optarg == NULL) {
				debug++;
			} else {
				if ((val = strtol(optarg, NULL, 0)) < 0)
					goto bad_option;
				debug = val;
			}
			break;
		case 'q':	/* -q, --quiet */
			if (debug)
				fprintf(stderr, "%s: suppressing normal output\n", argv[0]);
			debug = 0;
			output = 0;
			break;
		case 'v':	/* -v, --verbose [level] */
			if (debug)
				fprintf(stderr, "%s: increasing output verbosity\n", argv[0]);
			if (optarg == NULL) {
				output++;
				break;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			break;
		case 'h':	/* -h, --help */
		case 'H':	/* -H, --? */
			if (debug)
				fprintf(stderr, "%s: printing help message\n", argv[0]);
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			if (debug)
				fprintf(stderr, "%s: printing version message\n", argv[0]);
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			if (debug)
				fprintf(stderr, "%s: printing copying message\n", argv[0]);
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			optind--;
		      bad_nonopt:
			if (output || debug) {
				if (optind < argc) {
					fprintf(stderr, "%s: syntax error near '", argv[0]);
					while (optind < argc)
						fprintf(stderr, "%s ", argv[optind++]);
					fprintf(stderr, "'\n");
				} else {
					fprintf(stderr, "%s: missing option or argument", argv[0]);
					fprintf(stderr, "\n");
				}
				fflush(stderr);
			      bad_usage:
				usage(argc, argv);
			}
			exit(2);
		}
	}
	/* 
	 * dont' ignore non-option arguments
	 */
	if (optind < argc)
		goto bad_nonopt;
	strvf(argc, argv);
	exit(0);
}
